### DISCUSSION TITLE
Customising Click cli help with Rich

### DISCUSSION THREAD

**Discussion by ewels** [Textualize/rich#1121] [Ideas] [answered]
I feel like a common setup for `rich` users is to use the [`click` library](https://click.palletsprojects.com/) to build cli interfaces. The problem with `rich` making everything so pretty is that the `click` cli help texts start to stand out as being very... monochrome 😅 

I'm thinking that it should be possible to override the `click` formatter functions (see for example [this Stack Overflow question](https://stackoverflow.com/a/62437850/713980)) and was thinking of having a go at writing something as generic and reusable as possible that I could plug into several different projects.

Is this something that would be of interest to add to the `rich` package? Or is it too far off-scope? (I'm thinking it would be cool to have something similar to the `from rich.traceback import install` `install()` functionality). If you think it doesn't fit then I might whip up a tiny PyPI package for it, if I manage to make something half-decent.

Any tips or help from anyone very welcome! Also if anyone knows of something similar already in existence then that would of course be the best solution 💯

**Comment by Zethson** [comment]
Already answered you somewhere else, but reference implementations:

https://github.com/mlf-core/mlf-core/blob/master/mlf_core/custom_cli/click.py

https://github.com/cookiejar/cookietemple/blob/master/cookietemple/custom_cli/click.py

![image](https://user-images.githubusercontent.com/21954664/111871650-90ef6680-898b-11eb-8446-4903cda61b5b.png)

**Comment by willmcgugan** [comment, marked as answer]
Hi Phil,

I think it's a good idea! It would be nicer have to closer integration of the two libs, Typer would benefit as well.

I think something like this is better done as an external library. To keep dependencies down to a minimum, but also to reduce the amount of code for me to maintain. Perhaps it could be "rich-click" on Pypi. Or "click-rich".

Just saw @Zethson 's reply. A standalone lib with such functionality would be fantastic.

**Reply by Zethson** [reply]
Hi @willmcgugan,

this could be something that we aka @cookiejar might be able to do. We are currently rather busy, but maybe in a few months we could tackle this given that we already have a working solution.

CC @imipenem

**Reply by willmcgugan** [reply]
The would be great. If @ewels doesn't get there before you!

**Reply by willmcgugan** [reply]
Whoever tackles it, let me know if I can help!

**Reply by ewels** [reply]
Sounds good! I had a bit of a play the other day but didn't get very far.. I was mostly trying and failing to monkey patch `click` so that the package could be used with a single `install()` function instead of passing an argument to each click decorator. I was thinking that this would be good as it'd mean that people could keep their existing customisation functions.

@Zethson - the code you've written looks great! Would be nice to pull out a super generalised version of that which could be dropped into any project without modifying the `click` output too much (only the formatting). Let me know if you go down that route and I can certainly test + contribute.

**Reply by Zethson** [reply]
@ewels as mentioned, I won't have time anytime soon.
But maybe you want to have a crack at it since most of this code was written by you anyways, @Imipenem?

**Reply by Imipenem** [reply]
@Zethson sure, I can take a look at it maybe one or two weeks from now, around easter.

**Reply by ewels** [reply]
Ok, it's been nearly a year but once I saw what @willmcgugan had started with `rich-cli` I couldn't resist jumping in and ~stealing~ _adapting_ some of that beautiful help text rendering..

In case it's of any use to anyone, this is what I came up with: https://github.com/nf-core/tools/pull/1403 It's not nearly as complex as the stuff you guys have written @Zethson / @Imipenem but it is relatively generic now and expanded from @willmcgugan's code to also handle groups and few other cases.

@willmcgugan I'm sure that there is some cleaning up that could be done to the code here, but maybe another small step towards a package? What do you think? I would feel a bit weird about publishing this myself though, as it's mostly still your code 😁

**Reply by willmcgugan** [reply]
I think that looks amazing.

I’m not going to have the time to generalise the pretty help stuff in to a package. I’d be delighted if you wanted to tackle it. Let me know if you need any help, and feel free adapt as much as you like!

**Reply by ewels** [reply]
Alright then, here we go...

- GitHub: https://github.com/ewels/rich-click
- PyPI: https://pypi.org/project/rich-click/

I can't promise that it's well written, or even that it works 😅 But hopefully it's a starting point at least.

Super fun how little code is required with the monkey patching approach.. 😝

**Reply by abey79** [reply]
Looking great @ewels! Integrating this will be my first task after I push the current release of [my project](https://github.com/abey79/vpype). I'm hoping to contribute on the way. BTW, maybe you could enable Discussions for the project?

**Reply by ewels** [reply]
Great! Discussions enabled, see you there 🚀

**Reply by ssbarnea** [reply]
I am not sure exactly what @willmcgugan plans are regarding rich in general but I would like to suggest to welcome rich plugin developers to add their projects under the same organization as rich, following a pattern similar to other organizations: @tox-dev @sphinx-contrib @pytest-dev @apoclyps @click-contrib

I mention this because quote often after I started to use a plugin under a personal account I ended up with maintenance issues, the single point of failure issue.

For rich-click, there also the https://github.com/click-contrib is an option.

I seen that the "grouping" pattern, which is relatively new can really help building a closer community and fostering more collaboration -- tox and pytest being very good examples.

**Comment by ssbarnea** [comment]
Some time ago I created enrich package which adds few bits I found missing in rich. Feel free to propose a click/typer submodule to it, I use both and I am really interested about more integration.

As maintenance of any package is time consuming, it makes sense to team-up for tasks like this.

Some of the features may endup in rich, some not, but we can easily adapt and avoid cluttering rich itself with non mainstream stuff, especially as initially that new code would likely be... more experimental ;)

**Reply by willmcgugan** [reply]
Can't say I agree.  enrich has been where you implement stuff that I disagree with or think there is a better solution. Which is fine, but bundling that with other features would make them hard to discover.

**Reply by ssbarnea** [reply]
Indeed, from discoverability point of view separated is better.

Standalone extension ok too, I only wanted to help and mentioned that I am worried about those glue extensions can become unmaintained. I do still have some nightmares about some very useful sphinx extensions that ended being forked and re-released under different names just because original authors lost interest in maintaining them.

@willmcgugan as I seen that you do have an interest in click/typer, I can only expect to see a really nice integration in the future.

**Reply by ewels** [reply]
Wherever it ends up, more maintainers + repo collaborators etc = better 😄

**Comment by MLNW** [comment]
Some integration with click would be awesome!
